client/landing/stepper/declarative-flow/internals/hooks/use-sign-up-start-tracking/index.tsx

```diff
@@ -23,28 +24,35 @@ export const useSignUpStartTracking = ( { flow, currentStepRoute }: Props ) => {
 	const isFirstStep = firstStepSlug === currentStepRoute;
 	const flowVariant = flow.variantSlug;
 	const signupStartEventProps = flow.useSignupStartEventProps?.();
+	const shouldTrack = flow.isSignupFlow && ( isFirstStep || isSignupStep );
+	const flowName = flow.name;
 
-	const extraProps = useMemo(
-		() => ( {
+	const extraProps = useMemo( () => {
+		return {
 			...signupStartEventProps,
 			...( flowVariant && { flow_variant: flowVariant } ),
-		} ),
-		[ signupStartEventProps, flowVariant ]
-	);
-	const flowName = flow.name;
-	const shouldTrack = flow.isSignupFlow && ( isFirstStep || isSignupStep );
+		};
+	}, [ signupStartEventProps, flowVariant ] );
+
 	const removeSignupParam = useCallback( () => {
 		if ( queryParams.has( 'signup' ) ) {
 			queryParams.delete( 'signup' );
 			setQuery( queryParams );
 		}
 	}, [ queryParams, setQuery ] );
 
-	useEffect( () => {
-		if ( ! shouldTrack ) {
-			return;
-		}
-		recordSignupStart( flowName, ref, extraProps || {} );
+	const recordStart = useCallback( () => {
+		recordSignupStart( flowName, ref, extraProps );
 		removeSignupParam();
-	}, [ extraProps, flowName, ref, removeSignupParam, shouldTrack ] );
+	}, [ extraProps, flowName, ref, removeSignupParam ] );
+
+	useDisposableEffect(
```

Comment:
I am a bit confused here. How is this any different from ignoring the dependencies in the `useMemo`/`useCallback` calls above that would cause a rerender?

It's actually a little indirect if that's the case. We have the values being updated but falling in empty space on a "disposed" event (no effect). This feels cryptic. What am I missing?

Reply:
> I am a bit confused here. How is this any different from ignoring the dependencies in the useMemo/useCallback calls above that would cause a rerender?

@chriskmnds 

I have two goals with this PR.

1) Implement in a way that makes the code manageable for further evolutions.  
2) Express explicitly the intent of having this event triggered only once by 

Removing the dependency means that developers will never know if a problem was intentionally introduced or if it was a bug. Moreover, adding new external dependencies could easily bring back this issue. 
I suspect that this has already happened, as I am almost certain that the original implementation did not have the issues described.


>It's actually a little indirect if that's the case. We have the values being updated but falling in empty space on a "disposed" event (no effect). This feels cryptic. What am I missing?

It is exactly the code intent.

The original issue is we are expecting to have this event only on the being of flow and there are cases as you described in your [PR](https://github.com/Automattic/wp-calypso/pull/94174)  and for this specific event, we want to have it triggered only once by flow.

My understanding is that we can "accept"  having this event trigger multiple times in a specific case, but it is not 'OK' if we trigger this event when we user is in steps that are really not related to the signup flow. 

Let me express a use case.

Funnel: Analyzing the drop % of users who started the migration flow and dropped off before to see the checkout page.

So we have two steps to analyze

Step 1) Starting point 
calypso_signup_start {flow: migration }  

Step 2) Step migration_upgrade_plan

Step 3) checkout page.

 If, due to a code issue, the calypso_signup_start event is triggered after the checkout, the data in the funnel may become inaccurate. The flow will incorrectly register a user drop-off as a part of the flow usage.

Reply:
> Removing the dependency means that developers will never know if a problem was intentionally introduced or if it was a bug. Moreover, adding new external dependencies could easily bring back this issue.
I suspect that this has already happened, as I am almost certain that the original implementation did not have the issues described.

I'm not sure I agree with this reasoning. The idea of decentralizing the prop definitions (through hooks like `useEventProps`) is to eliminate the need to add more dependencies (or fiddle at all) with these framework-level hooks.

Not blocking the dependency, but treating it as blocked sounds like an anti-pattern to me. This isn't some generalised hook - it is precisely about "recording a Tracks event". I'm not arguing about ignoring or not the dependency - that's a different topic - just that the hook that's meant to "record the event" doesn't ignore the dependency at the point of recording the event. These callbacks just keep getting updated on every call for no reason or effect. A simple comment makes this abstraction completely redundant to me.

Ignoring the dependency raises other design concerns - such as why a dev would even be allowed to do something like:

```
useEventProps() {
  const { data } = useSiteIntent( 123 );
  return { 
    calypso_signup_start: {
      site_intent: data?.site_intent,
    } ,
  };
}
```

not knowing that this will most likely only record `undefined`. ETC. We discuss this separately though (let's not carry on here also).

Reply:
```
useEventProps() {
  const { data } = useSiteIntent( 123 );
  return { 
    calypso_signup_start: {
      site_intent: data?.site_intent,
    } ,
  };
}
`` 

Reactive code doesn't make much sense for this specific event because we can't wait until the site intent is loaded to trigger the signup_start and can't trigger it again when the site intent is loaded.   

OR

We evolve the approach to wait until the  calypso_signup_start has a value when it is defined on the `useEventProps` return. 


But I removed my abstraction ( and the comment) because it is an issue in production, and we need to merge it.  ae6e17ee320033cb8a3238c12a321957abb14325, we can evolve it in another opportunity.

Reply:
Thanks for keeping in wait. We have other discussion in https://github.com/Automattic/wp-calypso/issues/94175#issuecomment-2350500759, on which it looks like relaxing the criteria is possible, as analytics control what they observe. I'm going to reply there in a bit.

> Reactive code doesn't make much sense for this specific event

It's ok if "it doesn't make much sense". Flow implementations can be as "non-sensical" as they like. The question is not to block deeper analysis from happening where it needs to happen. 

> because we can't wait until the site intent is loaded to trigger the signup_start and can't trigger it again when the site intent is loaded

This is the restriction that I'm trying to probe about relaxing. Analytics can pick the first instance and that works irrespective.

> We evolve the approach to wait until the calypso_signup_start has a value when it is defined on the useEventProps return.

I think there is no difference between (1) the event triggered multiple times and analytics picking the first (or last), and (2) the event is not triggered until it has a value. In fact, the latter is also ambiguous, as that value may change at any point again. 

The focus, the way I see it, is that a specific flow implementation has included a given state variable to be recorded, and we want the system to respect that intention. Ignoring the dependency is akin to Stepper telling the consume (flow implementation): "You can do whatever you want through this hook/API, pass whatever props you want, async/sync everything goes, but I will just ignore". 

> Let me express a use case.
> 
> Funnel: Analyzing the drop % of users who started the migration flow and dropped off before to see the checkout page.
> 
> So we have two steps to analyze
> 
> Step 1) Starting point
> calypso_signup_start {flow: migration }
> 
> Step 2) Step migration_upgrade_plan
> 
> Step 3) checkout page.
> 
> If, due to a code issue, the calypso_signup_start event is triggered after the checkout, the data in the funnel may become inaccurate. The flow will incorrectly register a user drop-off as a part of the flow usage.

I'm trying to understand if this is indeed a problem. "After checkout" sounds like after flow unmounts, at which point everything is fresh mounted & rendered, so it doesn't sound relevant to ignoring or not a dependency. But also, like above, if analytics picks the first instance of this event, it won't even matter. 🤔

Reply:
The issue with checkout that I was pursuing was related to another issue fixed on #94364.  
After fixing that, I saw I was mixing contexts.